Support collections of executables in assertAll()

## Overview

Continuing the discussion that was spawned from this Twitter chat:
https://twitter.com/relizarov/status/930828269611401216

To start this discussion, these are the methods exposed by [AssertAll](https://github.com/junit-team/junit5/blob/master/junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertAll.java).

The `assertAll` method does not have an overload that takes a `Collection<Executable>`.

Currently to use `assertAll` from Kotlin you must first turn it into a Java8 `Stream<() -> Unit>`

This is the current example API:
```kotlin
@Test
fun `assertions from a stream`() {
    assertAll(
        "people with name starting with J",
        people
            .stream()
            .map {
                // This mapping returns Stream<() -> Unit>
                { assertTrue(it.firstName.startsWith("J")) }
            }
    )
}
```

I think what @elizarov was proposing in his tweet was to change the API so you could do something like this:

```kotlin
@Test
fun `assertions from a stream`() {
    assertAll(
        "people with name starting with J",
        people
            .map {
                // This mapping returns Stream<() -> Unit>
                { assertTrue(it.firstName.startsWith("J")) }
            }
    )
}
```

Currently the API exposes this method in kotlin:
```kotlin
/**
 * [Stream] of functions to be executed.
 */
private typealias ExecutableStream = Stream<() -> Unit>
private fun ExecutableStream.convert() = map { Executable(it) }

/**
 * @see Assertions.assertAll
 */
fun assertAll(executables: ExecutableStream) =
    Assertions.assertAll(executables.convert())

/**
 * @see Assertions.assertAll
 */
fun assertAll(heading: String?, executables: ExecutableStream) =
    Assertions.assertAll(heading, executables.convert())
```

The proposed suggestion would add additional assertion methods:
```kotlin
/**
 * [Collection] of functions to be executed.
 */
private typealias ExecutableCollection = Collection<() -> Unit>
/**
 * @see Assertions.assertAll
 */
fun assertAll(executables: ExecutableCollection) =
    assertAll(executables.stream())
/**
 * @see Assertions.assertAll
 */
fun assertAll(heading: String?, executables: ExecutableCollection) =
    assertAll(heading, executables.stream())
```

If the team decides that these methods make sense to be added, should be reflected in the Java API?
The additions would be something like this:

```java
static void assertAll(Collection<Executable> executables) {
    assertAll(null, executables);
}

static void assertAll(String heading, Collection<Executable> executables) {
    assertAll(heading, executables.stream());
}
```

## Question for the team

Why was `void assertAll(Collection<Executable> executables)` not added to the API to begin with?
Do those same reasons apply here? Do they only apply to Java but not Kotlin?

## Deliverables

- [x] Support `Collection<Executable>` in Java for `assertAll()`.
- [x] Document `Collection<Executable>` support in Java in Release Notes.
- [ ] Support `Collection<() -> Unit>` in Kotlin for `assertAll()`.
- [ ] Document `Collection<() -> Unit>` support in Kotlin in Release Notes (if necessary).
- [ ] Update Kotlin examples in User Guide.

Java Shell example works:
```
jshell
|  Welcome to JShell -- Version 9.0.1
|  For an introduction type: /help intro

jshell> void all(Runnable... args) { for(Runnable a : args) a.run(); }
|  created method all(Runnable...)

jshell> void all(Iterable<Runnable> iter) { for(Runnable a : iter) a.run(); }
|  created method all(Iterable<Runnable>)

jshell> all(() -> System.out.println("Hi"))
Hi

jshell> all(List.of(() -> System.out.println("Hi")))
Hi
```

 Sam tried it on master with the `assertAll` and one more tests failed:
```
/Users/sbrannen/source/junit5/junit-jupiter-engine/src/test/java/org/junit/jupiter/api/AssertAllAssertionsTests.java:133: error: unreported exception IOException; must be caught or declared to be thrown
                        throw new IOException();
```

--- a/documentation/src/test/kotlin/example/AssertionsDemoKotlin.kt
+++ b/documentation/src/test/kotlin/example/AssertionsDemoKotlin.kt
@@ -52,5 +52,13 @@ class AssertionsDemoKotlin {
         )
     }
 
+    @Test
+    fun `assertions from a collection`() {
+        assertAll(
+            "people with last name of Doe",
+            people.map { { assertEquals("Doe", it.lastName) } }
+        )
+    }
+
 }
 // end::user_guide[]

--- a/junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt
+++ b/junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt
@@ -35,6 +35,24 @@ fun assertAll(heading: String?, executables: ExecutableStream) =
     Assertions.assertAll(heading, executables.convert())
 
 /**
+ * [Collection] of functions to be executed.
+ */
+private typealias ExecutableCollection = Collection<() -> Unit>
+private fun ExecutableCollection.convert() = map { Executable(it) }
+
+/**
+ * @see Assertions.assertAll
+ */
+fun assertAll(executables: ExecutableCollection) =
+    Assertions.assertAll(executables.convert())
+
+/**
+ * @see Assertions.assertAll
+ */
+fun assertAll(heading: String?, executables: ExecutableCollection) =
+    Assertions.assertAll(heading, executables.convert())
+
+/**
  * @see Assertions.assertAll
  */
 fun assertAll(vararg executables: () -> Unit) =

